File: Tests/ZMLogTests.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import <XCTest/XCTest.h>
#import <ZMCSystem/ZMCSystem.h>
#import "ZMSLogging+Testing.h"

static char* const ZMLogTag = "Testing";
static NSString *const TagForTests = @"Testing";

@interface TestLog : NSObject

@property (nonatomic) NSString *tag;
@property (nonatomic) NSString *message;

- (instancetype)initWithTag:(NSString *)tag message:(NSString *)message;

@end



@implementation TestLog

- (instancetype)initWithTag:(NSString *)tag message:(NSString *)message
{
    self = [super init];
    if(self) {
        self.tag = tag;
        self.message = message;
    }
    return self;
}

@end


@interface ZMLogTests : XCTestCase

@property (nonatomic) NSMutableArray *receivedLogs; // array of TestLog

@end

@implementation ZMLogTests

- (void)setUp
{
    [super setUp];
    self.receivedLogs = [NSMutableArray array];
    ZMLogTestingResetLogLevels();
    ZMLoggingDebuggingHook = ^void(const char *tag, char const * const filename __unused, int linenumber __unused, NSString *output){
        [self.receivedLogs addObject:[[TestLog alloc] initWithTag:(tag ? [NSString stringWithUTF8String:tag] : nil) message:output]];
    };
}

- (void)tearDown
{
    [self removeTestLogSnapshotIfNeeded];
    ZMLogTestingResetLogLevels();
    [super tearDown];
}

- (void)removeTestLogSnapshotIfNeeded;
{
    NSFileManager *fm = [NSFileManager defaultManager];
    if ([fm fileExistsAtPath:[self filepathForSavingLogs]]) {
        [fm removeItemAtPath:[self filepathForSavingLogs] error:nil];
    }
}

- (NSString *)filepathForSavingLogs;
{
    
    NSError *error;
    NSFileManager *fm = [NSFileManager defaultManager];
    
    NSURL * const directory = [fm URLForDirectory:NSCachesDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:YES error:&error];
    RequireString(directory != nil, "Failed to get or create directory: %lu", (long) error.code);
    NSString *identifier = [NSBundle bundleForClass:[self class]].bundleIdentifier;
    NSURL *finalLocation = [[directory URLByAppendingPathComponent:identifier] URLByAppendingPathComponent:@"SavedLogs"];
    
    if (![[NSFileManager defaultManager] createDirectoryAtURL:finalLocation withIntermediateDirectories:YES attributes:nil error:&error]) {
        NSLog(@"error: %@", error);
    }
    NSString *filename = [NSString stringWithFormat:@"%@-test.log", identifier];
    return [finalLocation URLByAppendingPathComponent:filename].path;

}

- (void)testThatTheLogTagIsNotRegisteredIfNoLogIsCalled
{
    NSSet *tags = ZMLogGetAllTags();
    XCTAssertFalse([tags containsObject:TagForTests]);
}

- (void)testThatTheLogTagIsRegisteredAfterLoggingDebug
{
    ZMLogDebug(@"This log should cause the ZMLogTag to be registered");
    NSSet *tags = ZMLogGetAllTags();
    XCTAssertTrue([tags containsObject:TagForTests]);
}

- (void)testThatTheLogTagIsRegisteredAfterLoggingInfo
{
    ZMLogInfo(@"This log should cause the ZMLogTag to be registered");
    NSSet *tags = ZMLogGetAllTags();
    XCTAssertTrue([tags containsObject:TagForTests]);
}

- (void)testThatTheLoggingDebugHookIsNotCalledAfterLoggingInfo
{
    NSString *logMessage = @"This is a test message";
    ZMLogSetLevelForTag(ZMLogLevelInfo, TagForTests.UTF8String);
    ZMLogInfo(@"%@", logMessage);
    
    XCTAssertEqual(self.receivedLogs.count, 0u);
}

- (void)testThatTheLoggingDebugHookIsNotCalledAfterLoggingDebug
{
    NSString *logMessage = @"This is a test message";
    ZMLogDebug(@"%@", logMessage);
    
    XCTAssertEqual(self.receivedLogs.count, 0u);
}

- (void)testThatTheLoggingDebugHookIsCalledAfterLoggingWarning
{
    NSString *logMessage = @"This is a test message";
    ZMLogWarn(@"%@", logMessage);
    
    XCTAssertEqual(self.receivedLogs.count, 1u);
    TestLog *firstLog = self.receivedLogs.firstObject;
    XCTAssertNil(firstLog.tag);
    XCTAssertEqualObjects(firstLog.message, logMessage);
}

- (void)testThatTheLoggingDebugHookIsCalledAfterLoggingError
{
    NSString *logMessage = @"This is a test message";
    ZMLogError(@"%@", logMessage);
    
    XCTAssertEqual(self.receivedLogs.count, 1u);
    TestLog *firstLog = self.receivedLogs.firstObject;
    XCTAssertNil(firstLog.tag);
    XCTAssertEqualObjects(firstLog.message, logMessage);
}

- (void)testThatTheDefaultLogLevelIsWarning
{
    XCTAssertEqual(ZMLogGetLevelForTag("Testing"), ZMLogLevelWarn);
    XCTAssertEqual(ZMLogGetLevelForTag("FooBar"), ZMLogLevelWarn);
}

- (void)testThatTheLogLevelCanBeSet
{
    ZMLogSetLevelForTag(ZMLogLevelInfo, "FooBar");
    XCTAssertEqual(ZMLogGetLevelForTag("FooBar"), ZMLogLevelInfo);
}

- (void)testThatALogTagIsInitialized
{
    ZMLogInitForTag("Alice");
    ZMLogInitForTag("Bob");
    NSSet *expected = [NSSet setWithObjects:@"Bob", @"Alice", nil];
    XCTAssertEqualObjects(expected, ZMLogGetAllTags());
}

- (void)testThatLogMacrosCompile    
{
    ZMLogError(@"Test");
    ZMLogWarn(@"Test");
    ZMLogInfo(@"Test");
    ZMLogDebug(@"Test");
}

- (void)testThatSnapshotsIsFunctioning
{
    ZMLogDebug(@"value of i: %d", 1);
    ZMLogWarn(@"value of i: %d", 2);
    ZMLogError(@"value of i: %d", 3);
    
    NSString *filepath = [self filepathForSavingLogs];
    
    ZMLogSnapshot(filepath);
    XCTAssertTrue([[NSFileManager defaultManager] fileExistsAtPath:filepath]);
    NSString *logs = [NSString stringWithContentsOfFile:filepath encoding:NSASCIIStringEncoding error:nil];
    XCTAssertNotNil(logs);
}

@end

File: Tests/ZMLogTests.swift
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


import XCTest
@testable import ZMCSystem

class ZMLogTests: XCTestCase {

    override func setUp() {
        super.setUp()
        ZMSLog.debug_resetAllLevels();
    }
    
    override func tearDown() {
        ZMSLog.debug_resetAllLevels();
        ZMSLog.stopRecording()
        ZMSLog.removeAllLogHooks()
        super.tearDown()
    }
    
    func testThatTheLoggerRegistersATag() {
        
        // given
        let tag = "Foo"
        
        // when
        let _ = ZMSLog(tag: tag)
        
        // then
        let allTags = ZMSLog.allTags
        XCTAssertTrue(allTags.contains(tag))
    }
    
    func testThatTheLoggerRegistersTheRightLevel() {
        
        // given
        let tag = "Test11"
        
        // when
        let _ = ZMSLog(tag: tag)
        
        // then
        XCTAssertEqual(ZMSLog.getLevel(tag: tag), ZMLogLevel_t.warn)
    }
    
    func testThatTheLevelIsDebug() {
        
        // given
        let tag = "22test"
        let sut = ZMSLog(tag: tag)
        ZMSLog.set(level: .debug, tag: tag)
        
        // when
        var isDebug = false
        var isWarn = false
        var isInfo = false
        sut.ifDebug { isDebug = true }
        sut.ifInfo { isInfo = true }
        sut.ifWarn { isWarn = true }
        
        // then
        XCTAssertTrue(isDebug)
        XCTAssertTrue(isInfo)
        XCTAssertTrue(isWarn)
    }
    
    func testThatTheLevelIsWarn() {
        
        // given
        let tag = "22test"
        let sut = ZMSLog(tag: tag)
        ZMSLog.set(level: .warn, tag: tag)
        
        // when
        var isDebug = false
        var isWarn = false
        var isInfo = false
        sut.ifDebug { isDebug = true }
        sut.ifInfo { isInfo = true }
        sut.ifWarn { isWarn = true }
        
        // then
        XCTAssertFalse(isDebug)
        XCTAssertFalse(isInfo)
        XCTAssertTrue(isWarn)
    }
    
    func testThatTheLevelIsInfo() {
        
        // given
        let tag = "22test"
        let sut = ZMSLog(tag: tag)
        ZMSLog.set(level: .info, tag: tag)
        
        // when
        var isDebug = false
        var isWarn = false
        var isInfo = false
        sut.ifDebug { isDebug = true }
        sut.ifInfo { isInfo = true }
        sut.ifWarn { isWarn = true }
        
        // then
        XCTAssertFalse(isDebug)
        XCTAssertTrue(isInfo)
        XCTAssertTrue(isWarn)
    }
    
    func testThatTheLevelIsError() {
        
        // given
        let tag = "22test"
        let sut = ZMSLog(tag: tag)
        ZMSLog.set(level: .error, tag: tag)
        
        // when
        var isDebug = false
        var isWarn = false
        var isInfo = false
        sut.ifDebug { isDebug = true }
        sut.ifInfo { isInfo = true }
        sut.ifWarn { isWarn = true }
        
        // then
        XCTAssertFalse(isDebug)
        XCTAssertFalse(isInfo)
        XCTAssertFalse(isWarn)
    }
}

// MARK: - Log level management
extension ZMLogTests {
    
    func testThatLogIsNotRegisteredIfNoLogIsCalled() {
        XCTAssertEqual(ZMSLog.allTags.count, 0);
    }

    func testThatLogIsRegistered() {
        
        // GIVEN
        let tag = "Async"
        
        // WHEN
        ZMSLog.register(tag: tag)
        
        // THEN
        XCTAssertTrue(ZMSLog.allTags.contains(tag));
    }
    
    func testThatTheLogTagIsRegisteredAfterInitializingLog() {
        
        // GIVEN
        let tag = "Network"
        
        // WHEN
        let _ = ZMSLog(tag: tag)
        
        // THEN
        XCTAssertTrue(ZMSLog.allTags.contains(tag))
    }
    
    func testThatTheDefaultLogLevelIsWarning() {
        XCTAssertEqual(ZMSLog.getLevel(tag: "1234"), .warn)
    }

    func testThatTheLogLevelCanBeChanged() {
        
        // GIVEN
        let tag = "Draw"
        
        // WHEN
        ZMSLog.set(level: .debug, tag: tag)
        
        // THEN
        XCTAssertEqual(ZMSLog.getLevel(tag: tag), .debug)
    }
}

// MARK: - Debug hook
extension ZMLogTests {
    
    func testThatLogHookIsCalledWithError() {
        
        // GIVEN
        let tag = "Network"
        let level = ZMLogLevel_t.error
        let message = "PANIC!"
        
        let expectation = self.expectation(description: "Log received")
        let token = ZMSLog.addHook { (_level, _tag, _message) in
            XCTAssertEqual(level, _level)
            XCTAssertEqual(tag, _tag)
            XCTAssertEqual(message, _message)
            expectation.fulfill()
        }
        
        // WHEN
        ZMSLog(tag: tag).error(message)
        
        // THEN
        self.waitForExpectations(timeout: 0.5)
        
        // AFTER
        ZMSLog.removeLogHook(token: token)
    }
    
    func testThatLogHookIsNotCalledWithInfo() {
        
        // GIVEN
        let tag = "Network"
        let level = ZMLogLevel_t.info
        let message = "PANIC!"
        
        let token = ZMSLog.addHook { (_level, _tag, _message) in
            XCTAssertEqual(level, _level)
            XCTAssertEqual(tag, _tag)
            XCTAssertEqual(message, _message)
        }
        
        // WHEN
        ZMSLog(tag: tag).info(message)
        
        // THEN
        Thread.sleep(forTimeInterval: 0.2)
        
        // AFTER
        ZMSLog.removeLogHook(token: token)
    }
    
    func testThatLogHookIsCalledWithWarning() {
        
        // GIVEN
        let tag = "Network"
        let level = ZMLogLevel_t.warn
        let message = "PANIC!"
        
        let expectation = self.expectation(description: "Log received")
        let token = ZMSLog.addHook { (_level, _tag, _message) in
            XCTAssertEqual(level, _level)
            XCTAssertEqual(tag, _tag)
            XCTAssertEqual(message, _message)
            expectation.fulfill()
        }
        
        // WHEN
        ZMSLog(tag: tag).warn(message)
        
        // THEN
        self.waitForExpectations(timeout: 0.5)
        
        // AFTER
        ZMSLog.removeLogHook(token: token)
    }
    
    func testThatLogHookIsNotCalledWithDebug() {
        
        // GIVEN
        let tag = "Network"
        let level = ZMLogLevel_t.debug
        let message = "PANIC!"
        
        let token = ZMSLog.addHook { (_level, _tag, _message) in
            XCTAssertEqual(level, _level)
            XCTAssertEqual(tag, _tag)
            XCTAssertEqual(message, _message)
            XCTFail()
        }
        
        // WHEN
        ZMSLog(tag: tag).debug(message)
        
        // THEN
        Thread.sleep(forTimeInterval: 0.2)
        
        // AFTER
        ZMSLog.removeLogHook(token: token)
    }
    
    func testThatLogHookIsCalledWithDebugIfEnabled() {
        
        // GIVEN
        let tag = "Network"
        let level = ZMLogLevel_t.debug
        let message = "PANIC!"
        
        let expectation = self.expectation(description: "Log received")
        let token = ZMSLog.addHook { (_level, _tag, _message) in
            XCTAssertEqual(level, _level)
            XCTAssertEqual(tag, _tag)
            XCTAssertEqual(message, _message)
            expectation.fulfill()
        }
        
        // WHEN
        ZMSLog.set(level: .debug, tag: tag)
        ZMSLog(tag: tag).debug(message)
        
        // THEN
        self.waitForExpectations(timeout: 0.5)
        
        // AFTER
        ZMSLog.removeLogHook(token: token)
    }
    
    func testThatLogHookIsNotCalledWhenRemoved() {
        
        // GIVEN
        let tag = "Network"
        let message = "PANIC!"

        let token = ZMSLog.addHook { (_level, _tag, _message) in
            XCTFail()
        }
        ZMSLog.removeLogHook(token: token)
        
        // WHEN
        ZMSLog(tag: tag).error(message)
        Thread.sleep(forTimeInterval: 0.2)
    }
    
    
    func testThatLogHookIsNotCalledWhenRemovedAll() {
        
        // GIVEN
        let tag = "Network"
        let message = "PANIC!"
        
        let _ = ZMSLog.addHook { (_level, _tag, _message) in
            XCTFail()
        }
        ZMSLog.removeAllLogHooks()
        
        // WHEN
        ZMSLog(tag: tag).error(message)
        Thread.sleep(forTimeInterval: 0.2)
    }
    
    func testThatCallsMultipleLogHook() {
        
        // GIVEN
        let tag = "Network"
        let level = ZMLogLevel_t.error
        let message = "PANIC!"
        
        let expectation1 = self.expectation(description: "Log received")
        let expectation2 = self.expectation(description: "Log received")

        let token1 = ZMSLog.addHook { (_level, _tag, _message) in
            XCTAssertEqual(level, _level)
            XCTAssertEqual(tag, _tag)
            XCTAssertEqual(message, _message)
            expectation1.fulfill()
        }
        let token2 = ZMSLog.addHook { (_level, _tag, _message) in
            XCTAssertEqual(level, _level)
            XCTAssertEqual(tag, _tag)
            XCTAssertEqual(message, _message)
            expectation2.fulfill()
        }
        
        // WHEN
        ZMSLog(tag: tag).error(message)
        
        // THEN
        self.waitForExpectations(timeout: 0.5)
        
        // AFTER
        ZMSLog.removeLogHook(token: token1)
        ZMSLog.removeLogHook(token: token2)
    }
}

extension ZMLogTests {
    
    func testThatRecordedLogsAreEmptyWhenNotStarted() {
        
        // GIVEN
        let sut = ZMSLog(tag: "foo")
        
        // WHEN
        sut.error("PANIC")
        
        // THEN
        XCTAssertEqual(ZMSLog.recordedContent, [])
        
    }
    
    func testThatItRecordsLogs() {
        
        // GIVEN
        let sut = ZMSLog(tag: "foo")
        ZMSLog.startRecording()
        
        // WHEN
        sut.error("PANIC")
        sut.error("HELP")
        
        // THEN
        guard let logLine = ZMSLog.recordedContent.first else {
            XCTFail()
            return
        }
        XCTAssertTrue(logLine.hasSuffix("[0] [foo] PANIC"))
        XCTAssertEqual(ZMSLog.recordedContent.count, 2)
    }
    
    func testThatItDiscardsLogsWhenStopped() {
        
        // GIVEN
        let sut = ZMSLog(tag: "foo")
        ZMSLog.startRecording()
        
        // WHEN
        sut.error("PANIC")
        sut.error("HELP")
        ZMSLog.stopRecording()
        
        // THEN
        XCTAssertTrue(ZMSLog.recordedContent.isEmpty)
    }
}
